pkg/v1/partial/uncompressed.go

```diff
@@ -224,3 +277,58 @@ func (i *uncompressedImageExtender) LayerByDigest(h v1.Hash) (v1.Layer, error) {
 	}
 	return i.LayerByDiffID(diffID)
 }
+
+// compressedLayerFromUncompressedLayer implements partial.CompressedLayer
+type compressedLayerFromUncompressedLayer struct {
+	digest    v1.Hash
+	contents  []byte
+	mediaType types.MediaType
+}
+
+// ToCompressedLayer converts an uncompressed layer to a compressed layer
+func ToCompressedLayer(l v1.Layer) (v1.Layer, error) {
+	mediaType, err := l.MediaType()
+	if err != nil {
+		return nil, err
+	}
+	uncompressedReader, err := l.Uncompressed()
+	if err != nil {
+		return nil, err
+	}
+	gzipReadCloser := v1util.GzipReadCloser(uncompressedReader)
+
+	// TODO probably shouldnt cache the contents, should probably just re-gzip but aint nobody got time for that
+	buf := new(bytes.Buffer)
```

Comment:
I was thinking about users passing in a `io.ReadWriter` to allocate their own storage or something. Any feedback welcome.

Reply:
I'm not 100% sure what's going on here.

Reply:
This reads the compressed layer into RAM - which comes with tradeoffs. Could run out of RAM, but also get super fast layer contents access.